src/main/kotlin/com/github/jengelman/gradle/plugins/shadow/transformers/DeduplicatingResourceTransformer.kt

```diff
@@ -0,0 +1,135 @@
+package com.github.jengelman.gradle.plugins.shadow.transformers
+
+import java.io.File
+import java.security.MessageDigest
+import java.util.HexFormat
+import javax.inject.Inject
+import org.apache.tools.zip.ZipOutputStream
+import org.gradle.api.GradleException
+import org.gradle.api.file.FileTreeElement
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.tasks.Internal
+import org.gradle.api.tasks.util.PatternSet
+
+/**
+ * Transformer to include files with identical content only once in the shadow JAR.
+ *
+ * Multiple files with the same path but different content lead to an error.
+ *
+ * Some scenarios for duplicate resources in a shadow jar:
+ * * Duplicate `.class` files
+ *   Having duplicate `.class` files with different content is a situation indicating that the resulting jar is
+ *   built with _incompatible_ classes, likely leading to issues during runtime.
+ *   This situation can happen when one dependency is (also) included in an uber jar.
+ * * Duplicate `META-INF/<group-id>/<artifact-id>/pom.properties`/`xml` files.
+ *   Some dependencies contain shaded variants of other dependencies.
+ *   Tools that inspect jar files to extract the included dependencies, for example, for license auditing
+ *   use cases or tools that collect information of all included dependencies, may rely on these files.
+ *   Hence, it is desirable to retain the duplicate resource `pom.properties`/`xml` resources.
+ *
+ * `DeduplicatingResourceTransformer` checks all entries in the resulting jar.
+ * It is generally not recommended to use any of the [include] configuration functions.
+ *
+ * There are reasons to retain duplicate resources with different contents in the resulting jar.
+ * This can be achieved with the [exclude] configuration functions.
+ *
+ * To exclude a path or pattern from being deduplicated, for example, legit
+ * `META-INF/<group-id>/<artifact-id>/pom.properties`/`xml`, configure the transformer with an exclusion
+ * like the following:
+ * ```kotlin
+ * tasks.named<ShadowJar>("shadowJar").configure {
```

Comment:
```suggestion
 *
 * ```kotlin
 * tasks.shadowJar {
```